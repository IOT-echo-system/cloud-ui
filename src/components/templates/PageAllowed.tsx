import React, {useEffect, useState} from 'react'
import {useSelector} from '../../hooks'
import {useRouter} from 'next/router'
import {Config} from '../../config'
import {Loader} from '../atoms'
import {PolicyUtils} from '../../utils/policyUtils'

type PageAllowedPropsType<T extends Record<string, unknown>> = {
  policy: string
  Component: React.ComponentType<T>
} & T

export const PageAllowed = <T extends Record<string, unknown>>(props: PageAllowedPropsType<T>): React.JSX.Element => {
  const {policy, Component} = props
  const {policies} = useSelector(state => state.user)
  const [validated, setValidated] = useState(false)
  const router = useRouter()

  useEffect(() => {
    const isAllowed = PolicyUtils.isValid(policies, policy)
    if (!isAllowed) {
      router.push(Config.HOME_PAGE_PATH).catch(() => ({}))
    }
    setValidated(true)
  }, [])

  if (!validated) {
    return <Loader loadingText={'Loading...'} />
  }

  return <Component {...props} />
}
