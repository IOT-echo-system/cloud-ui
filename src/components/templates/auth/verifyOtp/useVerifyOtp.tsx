import type React from 'react'
import type {ChangeEvent} from 'react'
import {useState} from 'react'
import type {FormInputType} from '../../../atoms'
import type {CountDownTimerType} from '../../../../hooks'
import {useToast} from '../../../../hooks'
import {useCountDownTimer, useForm} from '../../../../hooks'
import {AuthService} from '../../../../services'
import type {GenerateOTPResBody, VerifyOTPResBody} from '../../../../services/typing/auth'
import {setStorage, StorageKeys} from '../../../../utils/storage'

type UseVerifyOtpReturnType = {
  inputFields: FormInputType[]
  handleSubmit: (event: React.FormEvent<HTMLFormElement>) => void
  otpGenerated: boolean
  countDownTimer: CountDownTimerType
  resendOtp: () => void
}
const useVerifyOtp = (otpVerifiedCallback: (status: boolean) => void): UseVerifyOtpReturnType => {
  const [otpGenerated, setOtpGenerated] = useState(false)
  const {values, onChange, handleSubmit} = useForm({email: '', otp: '', otpId: ''})
  const countDownTimer = useCountDownTimer(60)
  const toast = useToast()

  const handleChange = <K extends keyof typeof values>(keyName: K) => {
    return (event: ChangeEvent<HTMLInputElement>) => {
      onChange(keyName, event.target.value)
    }
  }

  const authorizeOTP = async (): Promise<GenerateOTPResBody | VerifyOTPResBody> => {
    if (otpGenerated) {
      const verifyOTPRes = await AuthService.verifyOTP(values)
      countDownTimer.pause()
      setStorage(StorageKeys.AUTH, {token: verifyOTPRes.token})
      otpVerifiedCallback(verifyOTPRes.success)
      return verifyOTPRes
    }
    const generateOTPRes = await AuthService.generateOTP(values.email)
    setOtpGenerated(true)
    onChange('otpId', generateOTPRes.otpId)
    countDownTimer.resetAndPlay()
    return generateOTPRes
  }

  const onSubmit = () => {
    authorizeOTP().catch(toast.error)
  }

  const resendOtp = (): void => {
    AuthService.generateOTP(values.email)
      .then(generateOTPRes => {
        setOtpGenerated(true)
        onChange('otpId', generateOTPRes.otpId)
        countDownTimer.resetAndPlay()
        return generateOTPRes
      })
      .catch(toast.error)
  }

  const inputFields: FormInputType[] = [
    {
      type: 'email',
      value: values.email,
      onChange: handleChange('email'),
      label: 'Email',
      required: true
    }
  ]

  const inputFieldsForOtp: FormInputType[] = [
    {
      type: 'email',
      value: values.email,
      onChange: handleChange('email'),
      label: 'Email',
      required: true,
      disabled: true
    },
    {
      type: 'number',
      value: values.otp,
      onChange: handleChange('otp'),
      label: 'Enter OTP',
      required: true
    }
  ]

  return {
    handleSubmit: handleSubmit(onSubmit),
    inputFields: otpGenerated ? inputFieldsForOtp : inputFields,
    countDownTimer,
    otpGenerated,
    resendOtp
  }
}

export {useVerifyOtp}
