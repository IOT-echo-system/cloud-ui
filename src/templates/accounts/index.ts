export * from './signup/SignUp'
export * from './login/Login'
