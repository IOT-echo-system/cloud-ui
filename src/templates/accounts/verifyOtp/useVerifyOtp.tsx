import type React from 'react'
import type {ChangeEvent} from 'react'
import {useState} from 'react'
import type {CountDownTimerType} from '../../../hooks'
import {useForm} from '../../../hooks'
import type {FormInputType} from '../../../atoms'
import AuthService from '../../../services/authService'
import type {ServerError} from '../../../typing/error'
import {useCountDownTimer} from '../../../hooks'
import {setStorage, StorageKeys} from '../../../utils/storage'
import type {GenerateOTPResBody, VerifyOTPResBody} from '../../../services/typing/auth'

type UseVerifyOtpReturnType = {
  inputFields: FormInputType[]
  handleSubmit: (event: React.FormEvent<HTMLFormElement>) => void
  error: string
  otpGenerated: boolean
  countDownTimer: CountDownTimerType
  resendOtp: () => void
}
const useVerifyOtp = (otpVerifiedCallback: (status: boolean) => void): UseVerifyOtpReturnType => {
  const [error, setError] = useState('')
  const [otpGenerated, setOtpGenerated] = useState(false)
  const {values, onChange, handleSubmit} = useForm({email: '', otp: '', otpId: ''})
  const countDownTimer = useCountDownTimer(60)

  const handleChange = <K extends keyof typeof values>(keyName: K) => {
    return (event: ChangeEvent<HTMLInputElement>) => {
      onChange(keyName, event.target.value)
    }
  }

  const authorizeOTP = async (): Promise<GenerateOTPResBody | VerifyOTPResBody> => {
    if (otpGenerated) {
      const verifyOTPRes = await AuthService.verifyOTP(values)
      countDownTimer.pause()
      setStorage(StorageKeys.AUTH, {token: verifyOTPRes.token})
      otpVerifiedCallback(verifyOTPRes.success)
      return verifyOTPRes
    }
    const generateOTPRes = await AuthService.generateOTP(values.email)
    setOtpGenerated(true)
    onChange('otpId', generateOTPRes.otpId)
    countDownTimer.resetAndPlay()
    return generateOTPRes
  }

  const onSubmit = () => {
    setError('')
    authorizeOTP().catch((error: ServerError) => {
      setError(error.message)
    })
  }

  const resendOtp = (): void => {
    AuthService.generateOTP(values.email)
      .then(generateOTPRes => {
        setOtpGenerated(true)
        onChange('otpId', generateOTPRes.otpId)
        countDownTimer.resetAndPlay()
        return generateOTPRes
      })
      .catch((error: ServerError) => {
        setError(error.message)
      })
  }

  const inputFields: FormInputType[] = [
    {
      type: 'email',
      value: values.email,
      onChange: handleChange('email'),
      label: 'Email',
      required: true
    }
  ]

  const inputFieldsForOtp: FormInputType[] = [
    {
      type: 'email',
      value: values.email,
      onChange: handleChange('email'),
      label: 'Email',
      required: true,
      disabled: true
    },
    {
      type: 'number',
      value: values.otp,
      onChange: handleChange('otp'),
      label: 'Enter OTP',
      required: true
    }
  ]

  return {
    error,
    handleSubmit: handleSubmit(onSubmit),
    inputFields: otpGenerated ? inputFieldsForOtp : inputFields,
    countDownTimer,
    otpGenerated,
    resendOtp
  }
}

export {useVerifyOtp}
