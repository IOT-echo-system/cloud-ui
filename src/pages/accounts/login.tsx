import type {NextPage} from 'next'
import React from 'react'
import {LogIn} from '@/templates/accounts'

const LoginPage: NextPage = () => {
  return <LogIn />
}

export default LoginPage
