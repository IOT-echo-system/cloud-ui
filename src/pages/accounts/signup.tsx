import type {NextPage} from 'next'
import React from 'react'
import {SignUp} from '@/templates/accounts'

const SignUpPage: NextPage = () => {
  return <SignUp />
}

export default SignUpPage
