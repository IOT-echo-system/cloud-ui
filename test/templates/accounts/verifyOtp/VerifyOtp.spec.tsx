import React from 'react'
import {render} from '@testing-library/react'
import * as UseVerifyOTPHook from '../../../../src/templates/accounts/verifyOtp/useVerifyOtp'
import {VerifyOtp} from '../../../../src/templates/accounts'

jest.mock('../../../../src/templates/accounts/verifyOtp/useVerifyOtp')

describe('Verify OTP Template Test', () => {
  const mockUseVerifyOTP = {
    inputFields: [{type: 'email', label: 'email'}],
    handleSubmit: jest.fn(),
    error: '',
    otpGenerated: false,
    countDownTimer: {
      time: 60,
      reset: jest.fn(),
      play: jest.fn(),
      resetAndPlay: jest.fn(),
      pause: jest.fn(),
      isRunning: jest.fn()
    },
    resendOtp: jest.fn()
  }

  beforeEach(() => {
    jest.resetAllMocks()
    jest.spyOn(UseVerifyOTPHook, 'useVerifyOtp').mockReturnValue(mockUseVerifyOTP)
  })
  afterEach(jest.resetAllMocks)

  it('should match with the snapshot', () => {
    const {container} = render(<VerifyOtp title={'Generate OTP'} setOtpVerified={jest.fn()} />)

    expect(UseVerifyOTPHook.useVerifyOtp).toHaveBeenCalledTimes(1)
    expect(container).toMatchSnapshot()
  })

  it('should match with the snapshot with otp generated', () => {
    const mockUseVerifyOTP = {
      inputFields: [{type: 'email', label: 'email'}],
      handleSubmit: jest.fn(),
      error: '',
      otpGenerated: true,
      countDownTimer: {
        time: 60,
        reset: jest.fn(),
        play: jest.fn(),
        resetAndPlay: jest.fn(),
        pause: jest.fn(),
        isRunning: jest.fn().mockReturnValue(true)
      },
      resendOtp: jest.fn()
    }
    jest.spyOn(UseVerifyOTPHook, 'useVerifyOtp').mockReturnValue(mockUseVerifyOTP)

    const {container} = render(<VerifyOtp title={'Verify OTP'} setOtpVerified={jest.fn()} />)

    expect(UseVerifyOTPHook.useVerifyOtp).toHaveBeenCalledTimes(1)
    expect(container).toMatchSnapshot()
  })

  it('should match with the snapshot with otp generated and after resend otp timer stop', () => {
    const mockUseVerifyOTP = {
      inputFields: [{type: 'email', label: 'email'}],
      handleSubmit: jest.fn(),
      error: '',
      otpGenerated: true,
      countDownTimer: {
        time: 60,
        reset: jest.fn(),
        play: jest.fn(),
        resetAndPlay: jest.fn(),
        pause: jest.fn(),
        isRunning: jest.fn().mockReturnValue(false)
      },
      resendOtp: jest.fn()
    }
    jest.spyOn(UseVerifyOTPHook, 'useVerifyOtp').mockReturnValue(mockUseVerifyOTP)

    const {container} = render(<VerifyOtp title={'Verify OTP'} setOtpVerified={jest.fn()} />)

    expect(UseVerifyOTPHook.useVerifyOtp).toHaveBeenCalledTimes(1)
    expect(container).toMatchSnapshot()
  })
})
