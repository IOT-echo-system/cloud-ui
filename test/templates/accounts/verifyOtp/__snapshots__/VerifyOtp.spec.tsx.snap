// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Verify OTP Template Test should match with the snapshot 1`] = `
<div>
  <form
    class="css-k6rfkb"
  >
    <div
      class="MuiStack-root css-1sazv7p-MuiStack-root"
    >
      <h5
        class="MuiTypography-root MuiTypography-h5 css-ag7rrr-MuiTypography-root"
      >
        Generate OTP
      </h5>
      <div
        class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="false"
          for=":r0:"
          id=":r0:-label"
        >
          email
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
            id=":r0:"
            type="email"
            value=""
          />
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-yjsfm1"
            >
              <span>
                email
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <div
        class="MuiStack-root css-pl8gqh-MuiStack-root"
      >
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeLarge MuiButton-containedSizeLarge MuiButton-fullWidth MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeLarge MuiButton-containedSizeLarge MuiButton-fullWidth css-uf1j15-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="submit"
        >
          Generate OTP
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </form>
</div>
`;

exports[`Verify OTP Template Test should match with the snapshot with otp generated 1`] = `
<div>
  <form
    class="css-k6rfkb"
  >
    <div
      class="MuiStack-root css-1sazv7p-MuiStack-root"
    >
      <h5
        class="MuiTypography-root MuiTypography-h5 css-ag7rrr-MuiTypography-root"
      >
        Verify OTP
      </h5>
      <div
        class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="false"
          for=":r1:"
          id=":r1:-label"
        >
          email
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
            id=":r1:"
            type="email"
            value=""
          />
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-yjsfm1"
            >
              <span>
                email
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <div
        class="MuiStack-root css-pl8gqh-MuiStack-root"
      >
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeLarge MuiButton-containedSizeLarge MuiButton-fullWidth Mui-disabled MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeLarge MuiButton-containedSizeLarge MuiButton-fullWidth css-uf1j15-MuiButtonBase-root-MuiButton-root"
          disabled=""
          tabindex="-1"
          type="button"
        >
          60
          s
          Resend OTP
        </button>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeLarge MuiButton-containedSizeLarge MuiButton-fullWidth MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeLarge MuiButton-containedSizeLarge MuiButton-fullWidth css-uf1j15-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="submit"
        >
          Verify OTP
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </form>
</div>
`;

exports[`Verify OTP Template Test should match with the snapshot with otp generated and after resend otp timer stop 1`] = `
<div>
  <form
    class="css-k6rfkb"
  >
    <div
      class="MuiStack-root css-1sazv7p-MuiStack-root"
    >
      <h5
        class="MuiTypography-root MuiTypography-h5 css-ag7rrr-MuiTypography-root"
      >
        Verify OTP
      </h5>
      <div
        class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="false"
          for=":r2:"
          id=":r2:-label"
        >
          email
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
            id=":r2:"
            type="email"
            value=""
          />
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-yjsfm1"
            >
              <span>
                email
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <div
        class="MuiStack-root css-pl8gqh-MuiStack-root"
      >
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeLarge MuiButton-outlinedSizeLarge MuiButton-fullWidth MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeLarge MuiButton-outlinedSizeLarge MuiButton-fullWidth css-h929h-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
        >
          Resend OTP
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeLarge MuiButton-containedSizeLarge MuiButton-fullWidth MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeLarge MuiButton-containedSizeLarge MuiButton-fullWidth css-uf1j15-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="submit"
        >
          Verify OTP
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </form>
</div>
`;
